core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -275,15 +277,9 @@ public DynamicResourcePool(Options options) {
         }
       }
     }, options.getInitialDelayMs(), options.getGcIntervalMs(), TimeUnit.MILLISECONDS);
-
-    registerGauges();
   }
 
-  private void registerGauges() {
-    MetricsSystem.registerGaugeIfAbsent(
-        MetricsSystem.getResourcePoolMetricName(this),
-        () -> size());
```

Comment:
This leaks a `mResources` reference and that reference leads to the references of all resources. Consequently all references that our Clients point to are left alive and not GC-ed.

Reply:
still not sure why a reference to `mResources` would cause the leak here, since the reference exist in the object pool class anyways? Have we verified that this change fixes the memory issue seen earlier? was it because of this line or the change in MetricSystem.java?

Reply:
![image](https://user-images.githubusercontent.com/14806853/163759858-9ba857f2-a388-4e17-83ca-49191fbc1171.png)
This is the ref chain. `() -> size()` effectively gives out `mResources.size()` which gives out `mResources`. After removing this gauge I no longer see the large memory consumption from `ConcurrentHashMap$Node` from the metrics registry. 

Bowen sent me this link earlier and seems to explain
https://blog.jooq.org/beware-of-functional-programming-in-java/

Reply:
I guess we can also say my current implementation is not the only possible fix, there are indeed other options:

1. By pooling, it is not easy to get rid of references to the things you pool. But maybe WeakReference can break the reference chain.

2. Indeed creating metrics per `FileSystemContext` with `DynamicResourcePool.registerGauges()` is too much excessive. And we are not de-registering those gauges. But I'm not even sure if HMS is caching FileSystemContext instances. Many compute engines may do that too, like Impala. I have not tested if only decreasing number of gauges will avoid the leak. However the thing is, if a client app has many `FileSystemContext`, what should a global gauge do(instead of being a counter)?

3. Maybe the reference chain can be broken at the client-connection step. But I haven't given it much thought.